### PR TITLE
Feature: add abacus toolchain

### PR DESCRIPTION
**Pull request by QuantumMisaka**
Toolchain scripts for quickly and properly compile and install dependencies of ABACUS including:

- gnu-toolchain
- - `gcc` (always in system)
- - `openblas`
- - `openmpi` or `mpich`
- - `scalapack`, `fftw`, `elpa` , `libxc`
- - `cereal`, `libnpy`, `libtorch`
- intel-toolchain (likely)

Version 2023.1 inspired by cp2k-toolchain.
Still need more test and improvement, but can have good usage.

The toolchain directory is designed to be in the path of ABACUS itself.

### CODE REVIEW

**Comment by dyzheng** [comment]
There are too much new files in this PR. Please check your changes.

**Comment by QuantumMisaka** [comment]
@dyzheng  Got it, much new file mainly in `libs/cereal` and `libs/libnpy`, now fix it.

**Review by caic99 — CHANGES_REQUESTED** [review]
@QuantumMisaka 
Thanks for your contribution! I believe these script will help users a lot.

However, please remove the release packages under libs. Doing so is an action of  redistribution and integration of those software packages. This is strongly not recommended because it might introduce incompatibility on licenses. (And you did not include the license files of those packages.)

It is OK to download them in your script. If you think integrating them in ABACUS repo is necessary, please use git submodule.
Thanks!

**Comment by caic99** [comment]
@QuantumMisaka And, if you wish, you can add docs or links to the toolchain under corresponding parts of `abacus-develop/docs` in online docs to promote it.

**Comment by QuantumMisaka** [comment]
@caic99 I also want to remove the libraries in `libs` and use toolchain to install it, I'm trying to do this but It's hard to choose the way to introduce the installation (by `git clone` or `wget` from the release, and neither way can bypass `github.com`, which will bring the HTTP problem). So I highly want to solve this with your help.

Also, the `docs` file is in plan. I'm wandering if the README file can be directly copy to `docs` directory, 

Thanks !

**Comment by caic99** [comment]
> @caic99 I also want to remove the libraries in `libs` and use toolchain to install it, I'm trying to do this but It's hard to choose the way to introduce the installation (by `git clone` or `wget` from the release, and neither way can bypass `github.com`, which will bring the HTTP problem). So I highly want to solve this with your help.

In principle it is the user's responsibiliy guaranteeing the internet connection. However, considering the actual situations, maybe you can provide some mirror sites of GitHub (e.g. fastgit) in place of the fixed base domain of github.com . Using [gitee mirroring service](https://gitee.com/mirrors) is a more secure way. 

> Also, the `docs` file is in plan. I'm wandering if the README file can be directly copy to `docs` directory

Please choose a good place for it to maintain the structure. Maybe under [quick_start](https://github.com/deepmodeling/abacus-develop/tree/develop/docs/quick_start) folder, and add an introduction for it at [easy_install.md](https://github.com/deepmodeling/abacus-develop/blob/develop/docs/quick_start/easy_install.md)? What do you think?

File: toolchain/libs/cereal-1.3.2.tar.gz
Comment (wenfei-li) [inline toolchain/libs/cereal-1.3.2.tar.gz:1]: I'm wondering if it is appropriate to include tar.gz file in the ABACUS repository

Comment (QuantumMisaka) [reply toolchain/libs/cereal-1.3.2.tar.gz:1]: It's surely not, I'm trying to add `cereal` and `libnpy` in toolchain installation

Comment (QuantumMisaka) [reply toolchain/libs/cereal-1.3.2.tar.gz:1]: Fixed it

File: toolchain/README.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+or link libraries ABACUS depends on 
+in ONLINE or OFFLINE way,
+and give setup files that you can use to compile ABACUS.
+
+## Todo
+- [x] `gnu-openblas` toolchain support for `openmpi` and `mpich`.
+- [x] `intel-mkl-mpi` toolchain support (for old oneapi using `icc` instead of `icx`).
+- [x] `intelmkl-openmpi/mpich` toolchain support (need more test).
+- [x] Automatic installation of [CEREAL](https://github.com/USCiLab/cereal) and [LIBNPY](https://github.com/llohse/libnpy) (by github.com)
+- [ ] A better mirror station for all packages, espeacially for CEREAL and LIBNPY.
```
Comment (caic99) [inline toolchain/README.md]:
```suggestion
- [ ] A better mirror station for all packages, especially for CEREAL and LIBNPY.
```

**Comment by QuantumMisaka** [comment]
@caic99 A problem in my try for introducing `LibRI` in `dep` to toolchain installation by using intel-icc-toolchain
```bash
In file included from /lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Tensor.h(8),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_lcao/hamilt_lcaodft/LCAO_matrix.h(13),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_io/restart.h(6),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_pw/hamilt_pwdft/global.h(8),
                 from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp(1):
/lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Shared_Vector.h(8): catastrophic error: cannot open source file "cereal/cereal.hpp"
  #include <cereal/cereal.hpp>

compilation aborted for /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp (code 4)
gmake[2]: *** [source/module_cell/CMakeFiles/cell.dir/build.make:244: source/module_cell/CMakeFiles/cell.dir/klist.cpp.o] Error 4
```
What's the best way for `LibRI` to find `CEREAL` ?

**Comment by caic99** [comment]
> @caic99 A problem in my try for introducing `LibRI` in `dep` to toolchain installation by using intel-icc-toolchain
> 
> ```shell
> In file included from /lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Tensor.h(8),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_lcao/hamilt_lcaodft/LCAO_matrix.h(13),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_io/restart.h(6),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_hamilt_pw/hamilt_pwdft/global.h(8),
>                  from /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp(1):
> /lustre/home/2201110432/apps/abacus/icc-mkl/deps/LibRI/include/RI/global/Shared_Vector.h(8): catastrophic error: cannot open source file "cereal/cereal.hpp"
>   #include <cereal/cereal.hpp>
> 
> compilation aborted for /lustre/home/2201110432/apps/abacus/icc-mkl/source/module_cell/klist.cpp (code 4)
> gmake[2]: *** [source/module_cell/CMakeFiles/cell.dir/build.make:244: source/module_cell/CMakeFiles/cell.dir/klist.cpp.o] Error 4
> ```
> 
> What's the best way for `LibRI` to find `CEREAL` ?

Hi @QuantumMisaka ,
Please check if `LibRI` has `CEREAL` for its git submodule. If not so, you can add `CEREAL` as an submodule in your toolchain.

**Comment by caic99** [comment]
> Hi @QuantumMisaka , Please check if `LibRI` has `CEREAL` for its git submodule. If not so, you can add `CEREAL` as an submodule in your toolchain.

@QuantumMisaka Update: Please check in your terminal if `cereal/cereal.hpp` can be opened in editor like `vim`. If the file is not present, the warning message would be `fatal error: xxx.h: No such file or directory`, but not `catastrophic error: cannot open source file`. The latter means the file presents, but the compiler can not open it. Please check file access permission.

**Comment by QuantumMisaka** [comment]
> @QuantumMisaka Update: Please check in your terminal if `cereal/cereal.hpp` can be opened in editor like `vim`. If the file is not present, the warning message would be `fatal error: xxx.h: No such file or directory`, but not `catastrophic error: cannot open source file`. The latter means the file presents, but the compiler can not open it. Please check file access permission.

- The permission of `cereal.hpp` is 644, which can be open by `vim`, 
- The problem still exists after change permission of `cereal.hpp` to `777`
- The problem will be fixed if `export CPATH="cereal/include":$CPATH

**Review by caic99 — APPROVED** [review]

**Pull request merged**